File: Algorithm/BOJ/Graph/Dijkstra/1753.최단경로.py
import heapq as hq

v, e = map(int, input().split())

inf = 10000000000

k = int(input())

a = [[] for _ in range(v+1)]

for _ in range(e):
    x, y, z = map(int, input().split())
    a[x].append((y, z))

dist = [inf]*(v+1)
check = [False]*(v+1)

dist[k] = 0

q = []

hq.heappush(q, (dist[k], k))

# heapq는 
# 야 거기 정점까지 갔어
# 거기 정점까지 최소 얼마들어
# 다음에 이거 빼서 점검해 이느낌

while q:
    _, x = hq.heappop(q)

    if check[x]: continue
    check[x] = True

    for ai in a[x]:
        y, cost = ai[0], ai[1]

        if dist[y] > dist[x] + cost:
            dist[y] = dist[x] + cost
            hq.heappush(q, (dist[y], y))
            
    
for i in range(1, v+1):
    if dist[i] == inf:
        print("INF")
    else:
        print(dist[i])

File: Algorithm/Advanced/Graph/Dijkstra/1753.최단경로.java
import java.util.*;
class Edge implements Comparable<Edge> {
    int to, cost;
    Edge(int to, int cost) {
        this.to = to;
        this.cost = cost;
    }
    public int compareTo(Edge that) {
        if (this.cost < that.cost) {
            return -1;
        } else if (this.cost == that.cost) {
            if (this.to < that.to) return -1;
            else if (this.to > that.to) return 1;
            else return 0;
        } else {
            return 1;
        }
    }
}
public class Main {
    static final int inf = 1000000000;
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        List<Edge>[] a = (List<Edge>[]) new List[n+1];
        for (int i=1; i<=n; i++) {
            a[i] = new ArrayList<Edge>();
        }
        int m = sc.nextInt();
        int start = sc.nextInt();
        for (int i=0; i<m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            int z = sc.nextInt();
            a[x].add(new Edge(y,z));
        }
        int[] dist = new int[n+1];
        boolean[] check = new boolean[n+1];
        for (int i=1; i<=n; i++) {
            dist[i] = inf;
            check[i] = false;
        }
        dist[start] = 0;
        PriorityQueue<Edge> q = new PriorityQueue<>();
        q.add(new Edge(start, 0));
        while (!q.isEmpty()) {
            int x = q.remove().to;
            if (check[x]) continue;
            check[x] = true;
            for (Edge y : a[x]) {
                if (dist[y.to] > dist[x] + y.cost) {
                    dist[y.to] = dist[x] + y.cost;
                    q.add(new Edge(y.to, dist[y.to]));
                }
            }
        }
        for (int i=1; i<=n; i++) {
            if (dist[i] == inf) {
                System.out.println("INF");
            } else {
                System.out.println(dist[i]);
            }
        }
    }
}